Flutter_TDD_Architecture_Course_notes.md

```diff
@@ -277,3 +277,310 @@ class GetRandomNumberTrivia extends UseCase<NumberTrivia, NoParams> {
+
+```dart 
+group('toJson', () {
+  test(
+    'should return a JSON map containing the proper data',
+    () async {
+      // act
+      final result = tNumberTriviaModel.toJson();
+      // assert
+      final expectedJsonMap = {
+        "text": "Test Text",
+        "number": 1,
+      };
+      expect(result, expectedJsonMap);
+    },
+  );
+});
+```
+Implement the method in the file `number_trivia_model.dart`.
+
+```dart
+class NumberTriviaModel extends NumberTrivia {
+  ...
+  Map<String, dynamic> toJson() {
+    return {
+      'text': text,
+      'number': number,
+    };
+  }
+}
+```
+Now all the tests **pass**.
+
+#### 5. Contracts of Data Sources
+> Video 5: https://www.youtube.com/watch?v=m_lkZo6CYcs&t=1s
+
+In the `fifth` tutorial an explanation is given about the importance of `contracts`.
+The contracts will allow the files to communicate directly with the **repository**.
+
+Inside the `repositories` folder create `number_trivia_repository_impl.dart`.
+This is the code for the file.
+
+```dart
+import 'package:dartz/dartz.dart';
+
+import '../../../../core/error/failure.dart';
+import '../../domain/entities/number_trivia.dart';
+import '../../domain/repositories/number_trivia_repository.dart';
+
+class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
+  @override
+  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
+    // TODO: implement getConcreteNumberTrivia
+    return null;
+  }
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
+    // TODO: implement getRandomNumberTrivia
+    return null;
+  }
+}
+```
+
+Inside the `platform` folder create a file called `network_info.dart`with the following code.
+
+```dart
+abstract class NetworkInfo {
+  Future<bool> get isConnected;
+}
+```
+#### Remote Data Source
+
+In this class the value returned will be different from the others and will return only a `NumberTriviaModel` value and the errors will be treated as `Exceptions`.
+
+Then we create the file `number_trivia_remote_data_source.dart`,with the following code.
+
+```dart
+import '../models/number_trivia_model.dart';
+
+abstract class NumberTriviaRemoteDataSource {
+  /// Calls the http://numbersapi.com/{number} endpoint.
+  ///
+  /// Throws a [ServerException] for all error codes.
+  Future<NumberTriviaModel> getConcreteNumberTrivia(int number);
+
+  /// Calls the http://numbersapi.com/random endpoint.
+  ///
+  /// Throws a [ServerException] for all error codes.
+  Future<NumberTriviaModel> getRandomNumberTrivia();
+}
+
+```
+#### Exceptions And Failures
+
+Whenever we don't have a number to show it will be a `ServerException`, then we should create it to be used.
+Inside the error folder create the file `exception.dart`.
+
+Code for `Exception.dart`.
+
+```dart
+class ServerException implements Exception {}
+
+class CacheException implements Exception {}
+```
+This is the code for the `failure.dart` file that also has to be created.
```

Comment:
_Where_ does the `failure.dart` file need to be created? Don't skip information. 💭